Prevent test coverage drop with automatic check

okurz opened the pull request:
Prevent test coverage drop with automatic check

The Makefile now features separate target to execute tests and gather coverage
data plus explicit target names for the report generation in 'coveralls'
format for travis CI and coveralls checks and 'html' format which we can use
for displaying locally.

We can now also call the new script 'check_coverage' which parses the HTML
coverage report for the 'total' coverage. The check fails with an error
message if the coverage drops below the threshold defined in the Makefile.

This way we can prevent further unintended coverage drop while allowing to
adjust the threshold when necessary. However we should try to increase the
threshold over time which could also be done by marking "uncoverable"
statements.

Fixes https://progress.opensuse.org/issues/10262

~~Merge this if you also approve #512 to supersede, else #512 first.~~

Verification by travis running the adjusted make targets itself:
https://travis-ci.org/os-autoinst/openQA/builds/105931851

aaannz commented:
Coverity itself has an option to set treshold when should coverity test report failure. I think we should first try that.

aaannz commented:
For now I set it to yous 59.4% Btw. it allows to even set max. allowed coverity decrease. Question now is if it actually works when comments aren't.

okurz commented:
@aaannz I don't know which "coverity" you mean here. "coveralls" is enabled and is also analyzing the builds but not publishing it in PRs as discussed in irc://chat.freenode.net/opensuse-factory . I would like developers to be able to locally reproduce the coverage analysis.

Using the make variable `COVERAGE_THRESHOLD` we can also set different values for travis checks, e.g. '0' to not make it fail when coverage drops but still keep the default for local testing. I would just recommend to stick to the threshold being the current level of coverage. As openQA is now deployed in at last three production environments we should care about what we change in openQA and therefore need a good coverage. *if* you still prefer travis checks to not change behaviour I would just call the make target from `.travis.yml` with `make COVERAGE_THRESHOLD=0 coverage-coveralls` for now

aaannz commented:
Of course I meant coveralls. I know they are enabled, but only to compute
and track coverage. There is option in coveralls to report failure when
coverage drops by some % or under some %. This was not enabled for openQA,
I now added yours treshold there. I don't know how it exactly works,
however. I simply don't think it's necessary to have custom result parsing
code when service is offering to do that automatically. (If it will work of
course, if it is the same issue as with comments, I rest my case)

2016-01-22 8:24 GMT+01:00 Oliver Kurz <notifications@github.com>:

> @aaannz <https://github.com/aaannz> I don't know which "coverity" you
> mean here. "coveralls" is enabled and is also analyzing the builds but not
> publishing it in PRs as discussed in irc://
> chat.freenode.net/opensuse-factory . I would like developers to be able
> to locally reproduce the coverage analysis.
>
> Using the make variable COVERAGE_THRESHOLD we can also set different
> values for travis checks, e.g. '0' to not make it fail when coverage drops
> but still keep the default for local testing. I would just recommend to
> stick to the threshold being the current level of coverage. As openQA is
> now deployed in at last three production environments we should care about
> what we change in openQA and therefore need a good coverage. *if* you
> still prefer travis checks to not change behaviour I would just call the
> make target from .travis.yml with make COVERAGE_THRESHOLD=0
> coverage-coveralls for now
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/os-autoinst/openQA/pull/513#issuecomment-173832052>.
>

okurz commented:
> I simply don't think it's necessary to have custom result parsing code when service is offering to do that automatically. (If it will work of course, if it is the same issue as with comments, I rest my case)

Maybe you prefer a simpler parsing of the text results and not HTML?

coolo commented:
OBS PRs have an extra coveralls check reporting failure correctly:
https://github.com/openSUSE/open-build-service/pull/1471

okurz commented:
updated:
 * rebased on current state in master

The diff should be much easier to look at. Parsing the text output instead of HTML is less reliable then looking at the well structured HTML output. I see coveralls checks work for OBS PRs but I can not easily see why they shouldn't for openQA, without being project admin even harder to do.

I like coveralls, it is giving good detailed reports. Nevertheless using the script I provide we have don't need to rely on coveralls as well as check locally. So everybody can easily check if the coverage threshold is still met. If one does not want to do it locally, travis/coveralls can still do it.
